common/src/main/java/io/netty/util/concurrent/Futures.java

```diff
@@ -0,0 +1,151 @@
+    }
+
+    static <A, B> void propagateUncommonCompletion(Future<? extends A> completed, Promise<B> recipient) {
+        if (completed.isCancelled()) {
+            tryCancel(recipient, LOGGER);
+        } else {
+            Throwable cause = completed.cause();
+            tryFailure(recipient, cause, LOGGER);
+        }
+    }
+
+    private static class PassThrough<R> implements FutureContextListener<Promise<R>, Object> {
+        @Override
+        public void operationComplete(Promise<R> recipient, Future<?> completed) throws Exception {
+            if (completed.isSuccess()) {
+                try {
+                    @SuppressWarnings("unchecked")
+                    R result = (R) completed.getNow();
+                    recipient.trySuccess(result);
+                } catch (RuntimeException e) {
+                    tryFailure(recipient, e, LOGGER);
+                }
+            } else {
+                propagateUncommonCompletion(completed, recipient);
+            }
+        }
+    }
+
+    private static class Mapper<R, T> implements FutureListener<Object> {
+        private final Promise<R> recipient;
+        private final Function<T, R> mapper;
+
+        Mapper(Promise<R> recipient, Function<T, R> mapper) {
+            this.recipient = recipient;
+            this.mapper = mapper;
+        }
+
+        @Override
+        public void operationComplete(Future<?> completed) throws Exception {
+            if (completed.isSuccess()) {
+                try {
+                    @SuppressWarnings("unchecked")
+                    T result = (T) completed.getNow();
+                    if (recipient.setUncancellable()) {
+                        R mapped = mapper.apply(result);
+                        recipient.setSuccess(mapped);
+                    }
+                } catch (RuntimeException e) {
+                    tryFailure(recipient, e, LOGGER);
+                }
+            } else {
+                propagateUncommonCompletion(completed, recipient);
+            }
+        }
+    }
+
+    private static class FlatMapper<R, T> implements FutureListener<Object> {
+        private final Promise<R> recipient;
+        private final Function<T, Future<R>> mapper;
+
+        FlatMapper(Promise<R> recipient, Function<T, Future<R>> mapper) {
+            this.recipient = recipient;
+            this.mapper = mapper;
+        }
+
+        @Override
+        public void operationComplete(Future<?> completed) throws Exception {
+            if (completed.isSuccess()) {
+                try {
+                    @SuppressWarnings("unchecked")
+                    T result = (T) completed.getNow();
+                    if (recipient.setUncancellable()) {
+                        Future<R> future = mapper.apply(result);
+                        future.addListener(recipient, passThrough());
+                        recipient.addListener(future, PromiseNotifier::propagateCancel);
```

Comment:
This may be applicable if you remove the `setUncancellable()` above but as it stands now, `recipient` can not be cancelled, so this listener is not required?

Reply:
I think the idea with `setUncancellable` isn't working out. Switching back to `trySuccess` but on the instance, not the logging one.